decompiler/pipeline/controlflowanalysis/variable_name_generation.py

```diff
@@ -72,26 +33,34 @@ class RenamingScheme(ABC):
     """Base class for different Renaming schemes."""
 
     def __init__(self, task: DecompilerTask) -> None:
-        """Collets all needed variables for renaming + filters already renamed + function arguments out"""
-        collector = VariableCollector(task._ast.condition_map)
+        """Collets all needed variables for renaming + filters which should not be renamed"""
+        collector = VariableCollector()
         collector.visit_ast(task._ast)
+        self._ast = task._ast
         self._params: List[Variable] = task._function_parameters
-        self._loop_vars : List[Variable] = collector.get_loop_variables()
-        self._variables: List[Variable] = list(filter(self._filter_variables, collector.get_variables()))
+        self._variables: Set[Variable] = set(filter(self._filter_variables, collector.get_variables()))
         
 
     def _filter_variables(self, item: Variable) -> bool:
         """Return False if variable is either a:
             - parameter
-            - renamed loop variable
             - GlobalVariable
         """
-        return not item in self._params and not (item in self._loop_vars and item.name.find("var_") == -1) and not isinstance(item, GlobalVariable)
+        return not item in self._params and not isinstance(item, GlobalVariable)
+
+
+    def renameVariables(self):
+        """Rename all collected variables with a naming scheme."""
+        names: dict[Variable, Variable] = {}
+        for var in self._variables:
+            names[var] = var.copy(self.getVariableName(var))
+
+        self._ast.replace_variables_in_subtree(self._ast.root, names)
```

Comment:
Are you sure you want to use the substitute here and not the replace function Hendrik is using?

My point is the following: 

The variables `var1 = Variable("var1", int, None, False, Variable("rax", int, 5), "some tag")` and variable `var2 = Variable("var1", int, None, False, Variable("rax", int, 7), "some other tag")` have the hash and are equivalent, although some values differ.

Now, if you find variable `var1` before variable `var2`, then each occurrence of `var2` is also replaced by variable `Variable("iVar3", int, None, False, Variable("rax", int, 5), "some tag")`.

There are two options, depending on what we want resp. need:
1. We say from this point on, we do not need the tags and old SSA-Variable names. Then we can ignore this information when creating the replacement variable, and both would be replaced by the same variable anyway.
2. We really only change the name, but then each variable need to be replaced by its unique new variable, i.e., we really check that we have the same object. 

In general, I would prefer the second, because we do not know whether we need it later, and it is more general. However, then I think your approach is not the most efficient one, because as soon as we find a variable, it would make sense to replace it in its expression.
However, when we decide that we do not need this information after this stage, you only have to update line 56, i.e., remove the unnecessary information when copying the variable.

Reply:
I get your point. 
The old algorithm, which simply renamed every variable (`var.name`), did it like in the second approach. 
But my problem was, how _messy_ variables are handled. 

I have __no grantee__ if a variable is a real copy of some other variable, or not. 
For example: Let's say we collected three times `var_0`, where the first and third are the same object and the second is a real copy. 

If I would iterate over every variable and simply assign the `.name` field, the third variable would change as I'm changing the first one. 
Afterwards, as the third one is being processed, if will again be renamed, destroying the name completely.
Using a set, does not help, because as you said, their hashes are the same.

The only other thing i could do, is to keep a list of `id`s of variables which I already renamed.

Reply:
And we are back to the point that pseudo should be immutable. 

Okay, but I think Hendrik did exactly this, considering the `id`s when replacing. So this is probably the cleanest way. Collecting variables according to their `id`s, assigning the new names and replace them with this new visitor. 

Seems very complicated for such an easy thing.

Reply:
Will look into his code/ask him today and probably block this request until his code is in main.

And __YES__, I __HATE IT__. 
Someone would think that renaming variables would be an easy task, but nope.
Hopefully with the id's everything works.

Reply:
Is this solved now and if so, where and how?

Reply:
The stage stores the new names as a dictionary and uses the `SubstituteVisitor` to iterate only once over each dataflow object.

1. Collection:
- `__init__()` of `HungarianScheme`
2. Iteration:
- `_rename_with_scheme` in the stage itself
3. Renaming:
- done via callback method `_rename_variable` of `HungarianScheme`

The AST visitor methods were removed.